core/src/main/scala/kafka/coordinator/transaction/TransactionManager.scala

```diff
@@ -0,0 +1,294 @@
+  private val corruptedPartitions: mutable.Set[Int] = mutable.Set()
+
+  /* Pid metadata cache indexed by transactional id, including ongoing transaction status */
+  private val pidMetadataCache = new Pool[String, PidMetadata]
+
+  /* shutting down flag */
+  private val shuttingDown = new AtomicBoolean(false)
+
+  def enablePidExpiration() {
+    scheduler.startup()
+
+    // TODO: add transaction and pid expiration logic
+  }
+
+  /**
+    * Get the pid metadata associated with the given groupId, or null if not found
+    */
+  def getPid(groupId: String): Option[PidMetadata] = {
+    Option(pidMetadataCache.get(groupId))
+  }
+
+  /**
+    * Add a new pid metadata, or retrieve the metadata if it already exists with the associated transactional id
+    */
+  def addPid(transactionalId: String, pidMetadata: PidMetadata): PidMetadata = {
+    val currentPidMetadata = pidMetadataCache.putIfNotExists(transactionalId, pidMetadata)
+    if (currentPidMetadata != null) {
+      currentPidMetadata
+    } else {
+      pidMetadata
+    }
+  }
+
+  def transactionTopicConfigs: Properties = {
+    val props = new Properties
+
+    props.put(LogConfig.UncleanLeaderElectionEnableProp, TransactionLog.EnforcedUncleanLeaderElectionEnable.toString)
+    props.put(LogConfig.CompressionTypeProp, TransactionLog.EnforcedCompressionCodec)
+    props.put(LogConfig.CleanupPolicyProp, TransactionLog.EnforcedCleanupPolicy)
+
+    props.put(LogConfig.MinInSyncReplicasProp, config.minInsyncReplicas.toString)
+    props.put(LogConfig.SegmentBytesProp, config.segmentBytes.toString)
+
+    props
+  }
+
+  def partitionFor(transactionalId: String): Int = Utils.abs(transactionalId.hashCode) % transactionTopicPartitionCount
+
+  def isCoordinatorFor(transactionalId: String): Boolean = inLock(partitionLock) {
+    val partitionId = partitionFor(transactionalId)
+
+    // partition id should be within the owned list and NOT in the corrupted list
+    ownedPartitions.contains(partitionId) && !corruptedPartitions.contains(partitionId)
+  }
+
+  /**
+    * Gets the partition count of the transaction log topic from ZooKeeper.
+    * If the topic does not exist, the default partition count is returned.
+    */
+  private def getTransactionTopicPartitionCount: Int = {
+    zkUtils.getTopicPartitionCount(Topic.TransactionStateTopicName).getOrElse(config.numPartitions)
+  }
+
+  private def loadPidMetadata(topicPartition: TopicPartition) {
+    def highWaterMark = replicaManager.getHighWatermark(topicPartition).getOrElse(-1L)
+
+    val startMs = time.milliseconds()
+    replicaManager.getLog(topicPartition) match {
+      case None =>
+        warn(s"Attempted to load offsets and group metadata from $topicPartition, but found no log")
+
+      case Some(log) =>
+        val buffer = ByteBuffer.allocate(config.loadBufferSize)
+
+        val loadedPids = mutable.Map.empty[String, PidMetadata]
```

Comment:
This is somehow inherited from the old two-type message thing, I think I can just merge `PidMetadata` and `TransactionMetadata` into `TransactionMetadata`, wdyt?